Dashboard: Support Variables in "Filter by Name" Transformation

Kirchen99 opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

This PR supports variables in "Filter by Name" transformation.

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/42267

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->


**Special notes for your reviewer**:

The `replace` function from the PR https://github.com/grafana/grafana/pull/44486 is used.

Discussion: https://github.com/grafana/grafana/discussions/50451

Kirchen99 committed:
add replace

Signed-off-by: Kirchen99 <latias_latios@126.com>

Kirchen99 committed:
check undefined

Signed-off-by: Kirchen99 <latias_latios@126.com>

Kirchen99 committed:
enable variable when using Filter by Name transformation

Signed-off-by: Kirchen99 <latias_latios@126.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=51804) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=51804) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=51804) it.</sub>

Kirchen99 committed:
consider clickable field in transform

Signed-off-by: Kirchen99 <latias_latios@126.com>

Kirchen99 committed:
simplify undefined checks

Signed-off-by: Kirchen99 <latias_latios@126.com>

JoaoSilvaGrafana commented on packages/grafana-data/src/transformations/transformers/filter.ts:
```diff
@@ -22,9 +22,25 @@ export const filterFieldsTransformer: DataTransformerInfo<FilterOptions> = {
+      return source.pipe(noopTransformer.operator({}, replace));
+    }
+
+    if (typeof options.include?.options === 'string') {
+      options.include.options = replace ? replace(options.include?.options) : options.include?.options;
+    } else if (typeof options.include?.options?.pattern === 'string') {
+      options.include.options.pattern = replace
+        ? replace(options.include?.options.pattern)
+        : options.include?.options.pattern;
+    }
+
+    if (typeof options.exclude?.options === 'string') {
+      options.exclude.options = replace ? replace(options.exclude?.options) : options.exclude?.options;
+    } else if (typeof options.exclude?.options?.pattern === 'string') {
+      options.exclude.options.pattern = replace
+        ? replace(options.exclude?.options.pattern)
+        : options.exclude?.options.pattern;
```
This is slightly hard to read, could we change it to have the `replace` as a condition at the top?
```
    if (replace) {
      if (typeof options.include?.options === 'string') {
        options.include.options = replace(options.include?.options);
      } else if (typeof options.include?.options?.pattern === 'string') {
        options.include.options.pattern = replace(options.include?.options.pattern);
      }

      if (typeof options.exclude?.options === 'string') {
        options.exclude.options = replace(options.exclude?.options);
      } else if (typeof options.exclude?.options?.pattern === 'string') {
        options.exclude.options.pattern = replace(options.exclude?.options.pattern);
      }
    }
```

JoaoSilvaGrafana reviewed (commented):
Hi @Kirchen99 👋 Thanks for submitting this PR!
I have left a comment on it, to make it a bit easier to understand. Also, could we add a test to both of these transformations that were changed, to test with a `replace`? You can check the test of `calculateField` for inspiration 😄 

PS: I have added a comment to this ticket https://github.com/grafana/grafana/issues/46428  to make sure we reimplement most of this using `getTemplateSrv` directly when we move the transformers.

Kirchen99 committed:
test: add test for filterByName

Signed-off-by: Kirchen99 <latias_latios@126.com>

Kirchen99 committed:
fix test lint

Signed-off-by: Kirchen99 <latias_latios@126.com>

Kirchen99 commented:
> Hi @Kirchen99 👋 Thanks for submitting this PR! I have left a comment on it, to make it a bit easier to understand. Also, could we add a test to both of these transformations that were changed, to test with a `replace`? You can check the test of `calculateField` for inspiration 😄
> 
> PS: I have added a comment to this ticket #46428 to make sure we reimplement most of this using `getTemplateSrv` directly when we move the transformers.

Please have a look if the test I added is sufficient.

JoaoSilvaGrafana reviewed (approved):
Looks good, thanks for submitting this! 🎉
